README.md
![FLOSS-logo](https://raw.githubusercontent.com/maliceio/malice-floss/master/logo.png)

malice-floss
============

[![Circle CI](https://circleci.com/gh/maliceio/malice-floss.png?style=shield)](https://circleci.com/gh/maliceio/malice-floss) [![License](http://img.shields.io/:license-mit-blue.svg)](http://doge.mit-license.org) [![Docker Stars](https://img.shields.io/docker/stars/malice/floss.svg)](https://hub.docker.com/r/malice/floss/) [![Docker Pulls](https://img.shields.io/docker/pulls/malice/floss.svg)](https://hub.docker.com/r/malice/floss/) [![Docker Image](https://img.shields.io/badge/docker image-114 MB-blue.svg)](https://hub.docker.com/r/malice/floss/)

Malice FLOSS Plugin

This repository contains a **Dockerfile** of the [FLOSS](https://github.com/fireeye/flare-floss) malice plugin **malice/floss**.

### Dependencies

-	[malice/alpine](https://hub.docker.com/r/malice/alpine/)

### Installation

1.	Install [Docker](https://www.docker.io/).
2.	Download [trusted build](https://hub.docker.com/r/malice/floss/) from public [DockerHub](https://hub.docker.com): `docker pull malice/floss`

### Usage

```
docker run --rm -v /path/to/file:/malware:ro malice/floss FILE
```

```bash
Usage: floss [OPTIONS] COMMAND [arg...]

Malice FLOSS Plugin

Version: v0.1.0, BuildTime: 20170130

Author:
  blacktop - <https://github.com/blacktop>

Options:
  --verbose, -V		    verbose output
  --timeout value       malice plugin timeout (in seconds) (default: 60) [$MALICE_TIMEOUT]
  --elasitcsearch value elasitcsearch address for Malice to store results [$MALICE_ELASTICSEARCH]
  --callback, -c	    POST results to Malice webhook [$MALICE_ENDPOINT]
  --proxy, -x		    proxy settings for Malice webhook endpoint [$MALICE_PROXY]
  --table, -t		    output as Markdown table
  --all, -a		        output ascii/utf-16 strings
  --help, -h		    show help
  --version, -v		    print the version

Commands:
  web   Create a FLOSS scan web service  
  help	Shows a list of commands or help for one command

Run 'floss COMMAND --help' for more information on a command.
```

This will output to stdout and POST to malice results API webhook endpoint.

Sample Output
-------------

### JSON:

```json
{
  "floss": {
    "decoded": [
      {
        "Location": "0x401047",
        "Strings": [
          "Ie_nkokbpAtep",
          "+^]g*dpi",
          "Ie_nkokbpD]ra=_g"
        ]
      },
      {
        "Location": "0x401059",
        "Strings": [
          "*lecnaC*",
          "Software\\Microsoft\\CurrentNetInf",
          "SYSTEM\\CurrentControlSet\\Control\\Lsa",
          "Software\\Microsoft\\Windows\\CurrentVersion\\Policies\\Explorer\\Run",
          "MicrosoftZj",
          "LhbqnrnesDwhs",
          "MicrosoftHaveExit",
          "LhbqnrnesG`ud@bj",
          "IEXPLORE.EXE",
          "/ver.htm",
          "/exe.htm",
          "/app.htm",
          "/myapp.htm",
          "/hostlist.htm",
          ".a`j-gsl",
          "/SomeUpList.htm",
          "/SomeUpVer.htm",
          "www.flyeagles.com",
          "www.km-nyc.com",
          "/restore",
          "/dizhi.gif",
          "/connect.gif",
          "\\$NtUninstallKB900727$",
          "\\netsvc.exe",
          "\\netscv.exe",
          "\\netsvcs.exe",
          "System Idle Process",
          "Program Files",
          "\\Internet Exp1orer",
          "forceguest",
          "AudioPort",
          "AudioPort.sys",
          "SYSTEM\\CurrentControlSet\\Services",
          "SYSTEM\\ControlSet001\\Services",
          "SYSTEM\\ControlSet002\\Services",
          "\\drivers\\",
          "\\DriverNum.dat"
        ]
      },
      {
        "Location": "0x40511A",
        "Strings": [
          "\\A|{@"
        ]
      },
      {
        "Location": "0x403DDA",
        "Strings": [
          ",pVA."
        ]
      },
      {
        "Location": "0x404DDE",
        "Strings": [
          "SMBs",
          "NTLMSSP",
          "Windows 2000 2195",
          "Windows 2000 5.0",
          "SMBr",
          "PC NETWORK PROGRAM 1.0",
          "LANMAN1.0",
          "Windows for Workgroups 3.1a",
          "LM1.2X002",
          "LANMAN2.1",
          "NT LM 0.12"
        ]
      }
    ],
    "stack": [
      "\\A|{@",
      "CAAA\\",
      "cmd.exe"
    ]
  }
}
```

### STDOUT (Markdown Table):

---

#### Floss

##### Decoded Strings

Location: `0x401047`
 - `Ie_nkokbpAtep`
 - `+^]g*dpi`
 - `Ie_nkokbpD]ra=_g`

Location: `0x402830`
 - `##########################################################################`

Location: `0x401059`
 - `*lecnaC*`
 - `Software\Microsoft\CurrentNetInf`
 - `SYSTEM\CurrentControlSet\Control\Lsa`
 - `Software\Microsoft\Windows\CurrentVersion\Policies\Explorer\Run`
 - `MicrosoftZj`
 - `LhbqnrnesDwhs`
 - `MicrosoftHaveExit`
 - `LhbqnrnesG`ud@bj`
 - `IEXPLORE.EXE`
 - `/ver.htm`
 - `/exe.htm`
 - `/app.htm`
 - `/myapp.htm`
 - `/hostlist.htm`
 - `.a`j-gsl`
 - `/SomeUpList.htm`
 - `/SomeUpVer.htm`
 - `www.flyeagles.com`
 - `www.km-nyc.com`
 - `/restore`
 - `/dizhi.gif`
 - `/connect.gif`
 - `\$NtUninstallKB900727$`
 - `\netsvc.exe`
 - `\netscv.exe`
 - `\netsvcs.exe`
 - `System Idle Process`
 - `Program Files`
 - `\Internet Exp1orer`
 - `forceguest`
 - `AudioPort`
 - `AudioPort.sys`
 - `SYSTEM\CurrentControlSet\Services`
 - `SYSTEM\ControlSet001\Services`
 - `SYSTEM\ControlSet002\Services`
 - `\drivers\`
 - `\DriverNum.dat`

Location: `0x40511A`
 - `\A|{@`

Location: `0x403DDA`
 - `,pVA.`

Location: `0x404DDE`
 - `SMBs`
 - `NTLMSSP`
 - `Windows 2000 2195`
 - `Windows 2000 5.0`
 - `SMBr`
 - `PC NETWORK PROGRAM 1.0`
 - `LANMAN1.0`
 - `Windows for Workgroups 3.1a`
 - `LM1.2X002`
 - `LANMAN2.1`
 - `NT LM 0.12`

##### Stack Strings

 - `\A|{@`
 - `CAAA\`
 - `cmd.exe`

---

Documentation
-------------

-	[To write results to ElasticSearch](https://github.com/maliceio/malice-floss/blob/master/docs/elasticsearch.md)
-	[To create a FLOSS scan micro-service](https://github.com/maliceio/malice-floss/blob/master/docs/web.md)
-	[To post results to a webhook](https://github.com/maliceio/malice-floss/blob/master/docs/callback.md)

### Issues

Find a bug? Want more features? Find something missing in the documentation? Let me know! Please don't hesitate to [file an issue](https://github.com/maliceio/malice-floss/issues/new)

### CHANGELOG

See [`CHANGELOG.md`](https://github.com/maliceio/malice-floss/blob/master/CHANGELOG.md)

### Contributing

[See all contributors on GitHub](https://github.com/maliceio/malice-floss/graphs/contributors).

Please update the [CHANGELOG.md](https://github.com/maliceio/malice-floss/blob/master/CHANGELOG.md) and submit a [Pull Request on GitHub](https://help.github.com/articles/using-pull-requests/).

### License

MIT Copyright (c) 2016-2017 **blacktop**


CHANGELOG.md
Change Log
==========

All notable changes to this project will be documented in this file.

[v0.1.0] - 2016-11-11
---------------------

### Fixed

### Added

-	Timeout to protect against bad files
- web service for file scans

### Removed

### Changed

-	Upgrade to new malice/alpine 3.5


docker-entrypoint.sh
#!/bin/dumb-init /bin/sh
set -e

# Note above that we run dumb-init as PID 1 in order to reap zombie processes
# as well as forward signals to all processes in its session. Normally, sh
# wouldn't do either of these functions so we'd leak zombies as well as do
# unclean termination of all our sub-processes.

# If the user is trying to run scan directly with some arguments, then
# pass them to virustotal.
if [ "${1:0:1}" = '-' ]; then
    set -- scan "$@"
fi

# Look for scan subcommands.
if [ "$1" = 'version' ]; then
    # This needs a special case because there's no help output.
    set -- virustotal "$@"
elif scan --help "$1" 2>&1 | grep -q "scan $1"; then
    # We can't use the return code to check for the existence of a subcommand, so
    # we have to use grep to look for a pattern in the help output.
    set -- scan "$@"
fi

# If we are running scan, make sure it executes as the proper user.
if [ "$1" = 'scan' ]; then
    set -- gosu malice "$@"
fi

exec "$@"


scan.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"context"

	log "github.com/Sirupsen/logrus"
	"github.com/fatih/structs"
	"github.com/gorilla/mux"
	"github.com/maliceio/go-plugin-utils/database/elasticsearch"
	"github.com/maliceio/go-plugin-utils/utils"
	"github.com/parnurzeal/gorequest"
	"github.com/urfave/cli"
)

// Version stores the plugin's version
var Version string

// BuildTime stores the plugin's build time
var BuildTime string

var path string

const (
	name     = "floss"
	category = "pe"
)

type pluginResults struct {
	ID   string      `json:"id" structs:"id,omitempty"`
	Data resultsData `json:"floss" structs:"floss"`
}

type floss struct {
	Results resultsData `json:"floss"`
}

type resultsData struct {
	ASCIIStrings   []string         `json:"ascii" structs:"ascii"`
	UTF16Strings   []string         `json:"utf-16" structs:"utf-16"`
	DecodedStrings []decodedStrings `json:"decoded" structs:"decoded"`
	StackStrings   []string         `json:"stack" structs:"stack"`
}

type decodedStrings struct {
	Location string   `json:"location" structs:"location"`
	Strings  []string `json:"strings" structs:"strings"`
}

func assert(err error) {
	if err != nil {
		log.WithFields(log.Fields{
			"plugin":   name,
			"category": category,
			"path":     path,
		}).Fatal(err)
	}
}

// scanFile scans file with all floss rules in the rules folder
func scanFile(timeout int, all bool) floss {

	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
	defer cancel()

	flossResults := floss{}
	// flossResults.Results = parseFlossOutput(RunCommand("./floss", "-g", path), all)
	output, err := utils.RunCommand(ctx, "/usr/bin/floss", "-g", path)
	assert(err)
	flossResults.Results = parseFlossOutput(output, err, all)

	return flossResults
}

func parseFlossOutput(flossOutput string, err error, all bool) resultsData {

	if err != nil {
		return resultsData{ASCIIStrings: []string{err.Error()}}
	}

	log.WithFields(log.Fields{
		"plugin":   name,
		"category": category,
		"path":     path,
	}).Debug("FLOSS Output: ", flossOutput)

	keepLines := []string{}
	results := resultsData{}
	var decodedStrArray []decodedStrings

	lines := strings.Split(flossOutput, "\n")
	// remove empty lines
	for _, line := range lines {
		if len(strings.TrimSpace(line)) != 0 {
			keepLines = append(keepLines, strings.TrimSpace(line))
		}
	}
	// build results data
	for i := 0; i < len(keepLines); i++ {
		if all {
			if strings.Contains(keepLines[i], "FLOSS static ASCII strings") {
				results.ASCIIStrings = utils.RemoveDuplicates(getASCIIStrings(keepLines[i+1 : len(keepLines)]))
			}
			if strings.Contains(keepLines[i], "FLOSS static UTF-16 strings") {
				results.UTF16Strings = utils.RemoveDuplicates(getUTF16Strings(keepLines[i+1 : len(keepLines)]))
			}
		}
		if strings.Contains(keepLines[i], "Decoding function at") {
			// get function location
			location, numOfStrings := getLocationAndNumOfDecodedStrs(keepLines[i])

			decodedStr := decodedStrings{
				Location: location,
				Strings:  utils.RemoveDuplicates(keepLines[i+1 : i+numOfStrings+1]),
			}
			decodedStrArray = append(decodedStrArray, decodedStr)
			i = i + numOfStrings
			continue
		} else if strings.Contains(keepLines[i], "stackstrings") {
			// get stackstrings
			line := strings.TrimPrefix(keepLines[i], "FLOSS extracted")
			numOfStrings, err := strconv.Atoi(strings.TrimSpace(strings.TrimSuffix(line, "stackstrings")))
			assert(err)
			// if len(keepLines) < i+numOfStrings+1 {
			results.StackStrings = keepLines[i+1 : i+numOfStrings+1]
			// } else {
			// 	log.Fatal("shiz went sideways.")
			// }
			i = i + numOfStrings
			continue
		}
	}
	results.DecodedStrings = decodedStrArray

	return results
}

func getLocationAndNumOfDecodedStrs(line string) (string, int) {
	numMatch := regexp.MustCompile("[0-9]+").FindAllString(line, -1)
	locMatch := regexp.MustCompile("0x[0-9A-F]+").FindAllString(line, -1)

	if len(locMatch) > 0 && len(numMatch) > 0 {
		num, err := strconv.Atoi(numMatch[len(numMatch)-1])
		assert(err)

		return locMatch[0], num
	}
	return "", 0
}

func getASCIIStrings(strArray []string) []string {
	asciiStrings := []string{}
	for _, str := range strArray {
		if strings.Contains(str, "FLOSS static UTF-16 strings") {
			break
		}
		asciiStrings = append(asciiStrings, str)
	}
	return asciiStrings
}

func getUTF16Strings(strArray []string) []string {
	utf16Strings := []string{}
	for _, str := range strArray {
		if strings.Contains(str, "FLOSS decoded") {
			break
		}
		utf16Strings = append(utf16Strings, str)
	}
	return utf16Strings
}

func printMarkDownTable(f floss, toString bool) string {

	if toString {
		return "STRING VERSION OF MARKDOWN TABLE"
	}

	fmt.Printf("#### Floss\n\n")
	if f.Results.ASCIIStrings != nil {
		fmt.Printf("##### ASCII Strings\n\n")
		for _, ascStr := range f.Results.ASCIIStrings {
			fmt.Printf(" - `%s`\n", ascStr)
		}
		fmt.Println()
	}
	if f.Results.UTF16Strings != nil {
		fmt.Printf("##### UTF-16 Strings\n\n")
		for _, utfStr := range f.Results.UTF16Strings {
			fmt.Printf(" - `%s`\n", utfStr)
		}
		fmt.Println()
	}
	fmt.Printf("##### Decoded Strings\n\n")
	if f.Results.DecodedStrings != nil {
		for _, decodedStr := range f.Results.DecodedStrings {
			fmt.Printf("Location: `%s`\n", decodedStr.Location)
			for _, dStr := range decodedStr.Strings {
				fmt.Printf(" - `%s`\n", dStr)
			}
			fmt.Println()
		}
	} else {
		fmt.Println(" - No Strings")
	}
	fmt.Printf("##### Stack Strings\n\n")
	if f.Results.StackStrings != nil {
		for _, stkStr := range f.Results.StackStrings {
			fmt.Printf(" - `%s`\n", stkStr)
		}
	} else {
		fmt.Println(" - No Strings")
	}
	return ""
}

func printStatus(resp gorequest.Response, body string, errs []error) {
	fmt.Println(body)
}

func webService() {
	router := mux.NewRouter().StrictSlash(true)
	router.HandleFunc("/scan", webAvScan).Methods("POST")
	log.Info("web service listening on port :3993")
	log.Fatal(http.ListenAndServe(":3993", router))
}

func webAvScan(w http.ResponseWriter, r *http.Request) {

	vars := mux.Vars(r)
	all := vars["all"]

	r.ParseMultipartForm(32 << 20)
	file, header, err := r.FormFile("malware")
	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprintln(w, "Please supply a valid file to scan.")
		log.Error(err)
	}
	defer file.Close()

	log.Debug("Uploaded fileName: ", header.Filename)

	tmpfile, err := ioutil.TempFile("/malware", "web_")
	if err != nil {
		log.Fatal(err)
	}
	defer os.Remove(tmpfile.Name()) // clean up

	data, err := ioutil.ReadAll(file)

	if _, err = tmpfile.Write(data); err != nil {
		log.Fatal(err)
	}
	if err = tmpfile.Close(); err != nil {
		log.Fatal(err)
	}

	// Do AV scan
	var fl floss
	path = tmpfile.Name()
	if strings.EqualFold(all, "true") || all != "" {
		fl = scanFile(60, true)
	} else {
		fl = scanFile(60, false)
	}

	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	w.WriteHeader(http.StatusOK)

	if err := json.NewEncoder(w).Encode(fl); err != nil {
		log.Fatal(err)
	}
}

func main() {

	var elastic string

	cli.AppHelpTemplate = utils.AppHelpTemplate
	app := cli.NewApp()

	app.Name = "floss"
	app.Author = "blacktop"
	app.Email = "https://github.com/blacktop"
	app.Version = Version + ", BuildTime: " + BuildTime
	app.Compiled, _ = time.Parse("20060102", BuildTime)
	app.Usage = "Malice FLOSS Plugin"
	app.ArgsUsage = "FILE to scan with FLOSS"
	app.Flags = []cli.Flag{
		cli.BoolFlag{
			Name:  "verbose, V",
			Usage: "verbose output",
		},
		cli.IntFlag{
			Name:   "timeout",
			Value:  120,
			Usage:  "malice plugin timeout (in seconds)",
			EnvVar: "MALICE_TIMEOUT",
		},
		cli.StringFlag{
			Name:        "elasitcsearch",
			Value:       "",
			Usage:       "elasitcsearch address for Malice to store results",
			EnvVar:      "MALICE_ELASTICSEARCH",
			Destination: &elastic,
		},
		cli.BoolFlag{
			Name:   "callback, c",
			Usage:  "POST results to Malice webhook",
			EnvVar: "MALICE_ENDPOINT",
		},
		cli.BoolFlag{
			Name:   "proxy, x",
			Usage:  "proxy settings for Malice webhook endpoint",
			EnvVar: "MALICE_PROXY",
		},
		cli.BoolFlag{
			Name:  "table, t",
			Usage: "output as Markdown table",
		},
		cli.BoolFlag{
			Name:  "all, a",
			Usage: "output ascii/utf-16 strings",
		},
	}
	app.Commands = []cli.Command{
		{
			Name:  "web",
			Usage: "Create a FLOSS scan web service",
			Action: func(c *cli.Context) error {
				webService()
				return nil
			},
		},
	}
	app.Action = func(c *cli.Context) error {

		var err error

		if c.Bool("verbose") {
			log.SetLevel(log.DebugLevel)
		}

		if c.Args().Present() {
			path, err = filepath.Abs(c.Args().First())
			assert(err)

			if _, err = os.Stat(path); os.IsNotExist(err) {
				assert(err)
			}

			floss := scanFile(c.Int("timeout"), c.Bool("all"))

			// upsert into Database
			elasticsearch.InitElasticSearch(elastic)
			elasticsearch.WritePluginResultsToDatabase(elasticsearch.PluginResults{
				ID:       utils.Getopt("MALICE_SCANID", utils.GetSHA256(path)),
				Name:     name,
				Category: category,
				Data:     structs.Map(floss.Results),
			})

			if c.Bool("table") {
				printMarkDownTable(floss, false)
			} else {
				flossJSON, err := json.Marshal(floss)
				assert(err)
				if c.Bool("post") {
					request := gorequest.New()
					if c.Bool("proxy") {
						request = gorequest.New().Proxy(os.Getenv("MALICE_PROXY"))
					}
					request.Post(os.Getenv("MALICE_ENDPOINT")).
						Set("X-Malice-ID", utils.Getopt("MALICE_SCANID", utils.GetSHA256(path))).
						Send(string(flossJSON)).
						End(printStatus)

					return nil
				}
				fmt.Println(string(flossJSON))
			}
		} else {
			log.Fatal(fmt.Errorf("Please supply a file to scan with FLOSS"))
		}
		return nil
	}

	err := app.Run(os.Args)
	assert(err)
}
